tests/vm/sut.go

```diff
@@ -0,0 +1,280 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package vm
+
+import (
+	"fmt"
+	"log"
+	"os"
+	"strconv"
+	"strings"
+	"syscall"
+	"time"
+
+	. "github.com/onsi/ginkgo/v2" //nolint:staticcheck,revive
+	. "github.com/onsi/gomega"    //nolint:staticcheck,revive
+
+	"github.com/pkg/errors"
+	"golang.org/x/crypto/ssh"
+)
+
+const (
+	Passive     = "passive"
+	Active      = "active"
+	Recovery    = "recovery"
+	LiveCD      = "liveCD"
+	UnknownBoot = "unknown"
+
+	TimeoutRawDiskTest = 600 // Timeout to connect for recovery_raw_disk_test
+)
+
+type SUT struct {
+	Host          string
+	Username      string
+	Password      string
+	SSHKey        []byte
+	Timeout       int
+	artifactsRepo string
+	TestVersion   string
+	CDLocation    string
+	MachineID     string
+	VMPid         int
+}
+
+func NewSUT() *SUT {
+	var (
+		err        error
+		user       string
+		sshKeyFile string
+		sshKey     []byte
+		pass       string
+		host       string
+		vmPid      int
+		timeout    = 180
+		value      int
+	)
+
+	if user = os.Getenv("SSH_USER"); user == "" {
+		user = "root"
+	}
+
+	if sshKeyFile = os.Getenv("SSH_KEY"); sshKeyFile == "" {
+		sshKeyFile = "../assets/testkey"
+	}
+
+	if sshKey, err = os.ReadFile(sshKeyFile); err != nil {
+		fmt.Printf("No ssh key file '%s' found!\n", sshKeyFile)
+	}
+
+	if pass = os.Getenv("SSH_PASS"); pass == "" {
+		pass = "uc0@linux"
+	}
+
+	if host = os.Getenv("SSH_HOST"); host == "" {
+		host = "127.0.0.1:2222"
+	}
+
+	vmPidStr := os.Getenv("VM_PID")
+	if value, err = strconv.Atoi(vmPidStr); err == nil {
+		By(fmt.Sprintf("Underlaying VM pid is set to: %d", value))
+		vmPid = value
+	}
+
+	valueStr := os.Getenv("SSH_TIMEOUT")
+	if value, err = strconv.Atoi(valueStr); err == nil {
+		timeout = value
+	}
+
+	return &SUT{
+		Host:          host,
+		Username:      user,
+		Password:      pass,
+		SSHKey:        sshKey,
+		MachineID:     "test",
+		Timeout:       timeout,
+		artifactsRepo: "",
+		CDLocation:    "",
+		VMPid:         vmPid,
+	}
+}
+
+// BootFrom returns the booting partition of the SUT
+// NOTE: passive/recovery are not handled yet!
+func (s *SUT) BootFrom() string {
+	out, err := s.command("cat /proc/cmdline")
+	ExpectWithOffset(1, err).ToNot(HaveOccurred())
+
+	switch {
+	case strings.Contains(out, "LABEL=SYSTEM"):
+		return Active
+	case strings.Contains(out, "live:CDLABEL"):
+		return LiveCD
+	default:
+		return UnknownBoot
+	}
+}
+
+func (s *SUT) EventuallyBootedFrom(image string) {
+	Eventually(func() error {
+		actual := s.BootFrom()
+		if actual != image {
+			return fmt.Errorf("expected boot from %s, actual %s", image, actual)
+		}
+
+		return nil
+	}, time.Duration(60)*time.Second, time.Duration(10)*time.Second).ShouldNot(HaveOccurred())
+}
+
+func (s *SUT) GetOSRelease(ss string) string {
```

Comment:
I'm conflicted. We perform assertions within the scope of these functions and in turn our tests look like they're not doing anything:
```
		It("Extract informations from booted OS", func() {
			osVersion := s.GetOSRelease("VERSION_ID")
			osArch := s.GetArch()

			By(fmt.Sprintf("OS Version: %s", osVersion))
			By(fmt.Sprintf("OS Architecture: %s", osArch))
		})
```

Can we refactor this so that SUT functions return values or errors and then adapt the tests to include these assertions?

Reply:
I fixed this a bit by checking the returned values. And For now it's a very simple test that needs to be extended and fixed. So I think that I will remove this specific block later.